client/executor/common/src/error.rs

```diff
@@ -25,92 +25,95 @@ use wasmi;
+	#[error("Runtime panicked: {0}")]
 	RuntimePanicked(String),
 	/// Invalid memory reference.
-	#[display(fmt="Invalid memory reference")]
+	#[error("Invalid memory reference")]
 	InvalidMemoryReference,
 	/// The runtime must provide a global named `__heap_base` of type i32 for specifying where the
 	/// allocator is allowed to place its data.
-	#[display(fmt="The runtime doesn't provide a global named `__heap_base`")]
+	#[error("The runtime doesn't provide a global named `__heap_base`")]
 	HeapBaseNotFoundOrInvalid,
 	/// The runtime WebAssembly module is not allowed to have the `start` function.
-	#[display(fmt="The runtime has the `start` function")]
+	#[error("The runtime has the `start` function")]
 	RuntimeHasStartFn,
 	/// Some other error occurred
+	#[error("Other: {0}")]
 	Other(String),
 	/// Some error occurred in the allocator
-	#[display(fmt="Error in allocator: {}", _0)]
-	Allocator(sp_allocator::Error),
+	#[error("Allocation Error")]
+	Allocator(#[from] sp_allocator::Error),
 	/// Execution of a host function failed.
-	#[display(fmt="Host function {} execution failed with: {}", _0, _1)]
+	#[error("Host function {0} execution failed with: {1}")]
 	FunctionExecution(String, String),
 	/// No table is present.
 	///
 	/// Call was requested that requires table but none was present in the instance.
-	#[display(fmt="No table exported by wasm blob")]
+	#[error("No table exported by wasm blob")]
 	NoTable,
 	/// No table entry is present.
 	///
 	/// Call was requested that requires specific entry in the table to be present.
-	#[display(fmt="No table entry with index {} in wasm blob exported table", _0)]
-	#[from(ignore)]
+	#[error("No table entry with index {0} in wasm blob exported table")]
 	NoTableEntryWithIndex(u32),
 	/// Table entry is not a function.
-	#[display(fmt="Table element with index {} is not a function in wasm blob exported table", _0)]
-	#[from(ignore)]
+	#[error("Table element with index {0} is not a function in wasm blob exported table")]
 	TableElementIsNotAFunction(u32),
 	/// Function in table is null and thus cannot be called.
-	#[display(fmt="Table entry with index {} in wasm blob is null", _0)]
-	#[from(ignore)]
+	#[error("Table entry with index {0} in wasm blob is null")]
 	FunctionRefIsNull(u32),
-}
 
-impl std::error::Error for Error {
-	fn source(&self) -> Option<&(dyn std::error::Error + 'static)> {
-		match self {
-			Error::InvalidData(ref err) => Some(err),
-			Error::Trap(ref err) => Some(err),
-			Error::Wasmi(ref err) => Some(err),
-			_ => None,
-		}
-	}
+	#[error(transparent)]
+	RuntimeConstruction(#[from] WasmError),
+	
+	#[error("Shared memory is not supported")]
+	SharedMemUnsupported,
+	
+	#[error("Imported globals are not supported yet")]
+	ImportedGlobalsUnsupported,
+	
+	#[error("initializer expression can have only up to 2 expressions in wasm 1.0")]
+	InitializerHasTooManyExpressions,
+	
+	#[error("Invalid initializer expression provided {0}")]
+	InvalidInitializerExpression(String),
```

Comment:
We added here a bunch of error variants, but kept the `Other(String)`. Could you elaborate on what is the plan wrt `Other(String)` error type, is it going to removed eventually? It's not clear when to stop adding more error types.

I'm starting to think we don't need to present the user all the error types, a typical use-case would not match on error variants, but will log them.

Reply:
My intention is to provide sufficient depth so the chain of errors is sufficiently detailed to do a diagnostic right away.

Reply:
I have no intention to dig any further any more of the `Other` rabbit holes at this point.